src/java/org/apache/cassandra/io/util/FileUtils.java

```diff
@@ -68,19 +71,29 @@
     private static final boolean canCleanDirectBuffers;
     private static final AtomicReference<FSErrorHandler> fsErrorHandler = new AtomicReference<>();
 
+    private static Class clsDirectBuffer;
+    private static MethodHandle mhDirectBufferCleaner;
+    private static MethodHandle mhCleanerClean;
+
     static
     {
         boolean canClean = false;
         try
         {
+            clsDirectBuffer = Class.forName("sun.nio.ch.DirectBuffer");
+            Method mDirectBufferCleaner = clsDirectBuffer.getMethod("cleaner");
+            mhDirectBufferCleaner = MethodHandles.lookup().unreflect(mDirectBufferCleaner);
+            Method mCleanerClean = mDirectBufferCleaner.getReturnType().getMethod("clean");
+            mhCleanerClean = MethodHandles.lookup().unreflect(mCleanerClean);
+
             ByteBuffer buf = ByteBuffer.allocateDirect(1);
-            ((DirectBuffer) buf).cleaner().clean();
```

Comment:
If we keep the `canClean` code, we should make sure we try and cleanup the above `buf`, and then also remove re-throwing the runtime exception below. As otherwise, we will always think we can cleanup memory mapped files, which may not be true (and I think we will fail randomly during execution). But, I could totally be reading this wrong :D

Reply:
The tests in linux show that the cleaner does indeed work with offheap-buffers.